Fix sample packing producing longer sequences than specified by `sequence_len`

tobmi1 opened the pull request:
# Description

Return tokenized sequence lengths in `get_dataset_lengths` directly from the `Dataset` object instead of accessing the underlying PyArrow table. Filtering the dataset (e.g., by applying `drop_long_seq_in_dataset`) would update the `Dataset` object but not the underlying table. Thus, the sequence lengths returned by `get_dataset_lengths` would no longer correspond to the correct sequences and the sample packing algorithm may produce samples longer than `cfg.sequence_len`.

## Motivation and Context

Training with sample packing and `pad_to_sequence_len=True` should result in stable VRAM usage. However, due to this bug, sequence length may differ between batches, which caused increased VRAM usage and even OOMs in the middle of training in my case. Fine-tuning Phi-3.5 with `sequence_len=4096`, `micro_batch_size=1`, `gradient_accumulation_steps=4`, and `gradient_checkpointing=False` went from random OOMs after ~500 steps to using stable ~42GB VRAM on a single A100 80GB with the fix.

Note that pretraining does not seem to be affected since the underlying in-memory table is updated along with the `IterableDataset` when applying any filters.

Pre-processing the dataset in a separate step and then loading the prepared dataset for training should theoretically also circumvent this issue, but I was not able confirm this yet.

## How has this been tested?

I extended `test_packed_batch_sampler.py` to use `drop_long_seq_in_dataset` before applying the `MultipackBatchSampler` and parametrized the test to additionally use a `sequence_len` of 512 so that some samples are actually dropped from the dataset. Running the extended test without the fix fails, while it runs successfully on my setup with the fix.

My env:
OS: Windows 10 (WSL Ubuntu 24.04)
Python: 3.12.9
Torch: 2.5.1 (CPU-only)

## Screenshots (if appropriate)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

## Social Handles (Optional)

<!-- Thanks for submitting a bugfix or enhancement. -->
<!-- We'd love to show our thanks to you on Twitter & Discord if you provide your handle -->

tobmi1 committed:
Extend MultiPackBatchSampler test to include shorter sequence length and drop long sequences filter

tobmi1 committed:
Fix get_dataset_lengths for datasets that were previously filtered (e.g., with drop_long_seq_in_dataset)

NanoCode012 reviewed (dismissed):
Thank you for catching this. I wonder whether this is the cause for those silent vram increases.

NanoCode012 commented:
Tests are failing with 

```
>           dataset, _ = load_tokenized_prepared_datasets(
                self.tokenizer, cfg, prepared_path
            )

tests/test_datasets.py:231: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/axolotl/utils/data/sft.py:325: in load_tokenized_prepared_datasets
    dataset = drop_long_seq_in_dataset(dataset, cfg)
src/axolotl/utils/data/utils.py:175: in drop_long_seq_in_dataset
    min_input_len = np.min(get_dataset_lengths(dataset))
src/axolotl/utils/samplers/utils.py:15: in get_dataset_lengths
    lengths = np.vectorize(len)(np.array(input_ids, dtype=object))

...

            inputs = [arg.flat[0] for arg in args]
>           outputs = func(*inputs)
E           TypeError: object of type 'int' has no len()
```

NanoCode012 commented on src/axolotl/utils/samplers/utils.py:
```diff
@@ -5,12 +5,12 @@
     else:
-        input_ids = dataset.data.column("input_ids")
+        input_ids = dataset["input_ids"]
         lengths = np.vectorize(len)(np.array(input_ids, dtype=object))
```
Maybe try this change? Untested!

```suggestion
        lengths = np.array([len(seq) for seq in input_ids])
```

Since np.vectorize is not for perf but convenience (https://numpy.org/doc/stable/reference/generated/numpy.vectorize.html) , it shouldn't be much different.

tobmi1 replied:
Thank you for the suggestion! The tests in `test_datasets.py` are now running fine on my machine. However, a bunch of others (e.g. in `tests/patched/test_validation.py`) are still failing for seemingly unrelated reasons. I hope this is just due to my env with WSL

tobmi1 committed:
Update src/axolotl/utils/samplers/utils.py

Fix get_dataset_lengths for datasets that do not have position_ids or length attributes

Co-authored-by: NanoCode012 <kevinvong@rocketmail.com>

NanoCode012 reviewed (approved):
Checked the test, confirmed error appears with prev code.
